File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/DocHeadingsRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires specific headings in documentation files.
 */
public enum DocHeadingsRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "docHeadings";
    }
    private static final String CATEGORY = "docHeadings";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final List<String> headings = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("headings"));
        final JsonNode sourceFilter = catNode.get("parameters").get("sourceFilter");
        final String prefix = sourceFilter.get("prefix").asText();
        final String suffix = sourceFilter.get("suffix").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(manifest.get("filePresence").get("parameters").get("paths")).stream()
                .filter(f -> f.startsWith(prefix))
                .filter(f -> f.endsWith(suffix))
                .flatMap(f -> validateHeadings(root, f, headings, severity).stream())
                .toList();
    }

    private List<Finding> validateHeadings(Path root, String file, List<String> headings, String severity) {
        final Path filePath = root.resolve(file);
        return Stream.of(filePath)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .map(p -> {
                    try {
                        return HarnessCheckHelper.readFile(root, p);
                    } catch (MojoExecutionException e) {
                        return "";
                    }
                })
                .flatMap(text -> headings.stream()
                        .filter(h -> !text.contains(h))
                        .map(h -> new Finding(severity, CATEGORY, "doc missing " + h + ": " + file)))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/DocHeadingsRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified headings.
 */
object DocHeadingsRule : HarnessCheckRule() {
    override val category: String = "docHeadings"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> {
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        return when {
            catObj == null || parametersObj == null -> {
                emptyList()
            }

            else -> {
                val sourceCategory = HarnessCheck.stringFrom(parametersObj, "sourceFilesFromCategory")
                val sourceFiles =
                    HarnessCheck.stringArrayFrom(
                        manifest[sourceCategory]?.jsonObject?.get("parameters")?.jsonObject,
                        "paths",
                    )
                val sourceFilterObj = parametersObj["sourceFilter"]?.jsonObject
                val prefix = HarnessCheck.stringFrom(sourceFilterObj, "prefix")
                val suffix = HarnessCheck.stringFrom(sourceFilterObj, "suffix")
                val headings = HarnessCheck.stringArrayFrom(parametersObj, "headings")
                sourceFiles
                    .filter { sourceFile -> sourceFile.startsWith(prefix) }
                    .filter { sourceFile -> sourceFile.endsWith(suffix) }
                    .flatMap { docPath ->
                        headings
                            .filter { heading ->
                                !HarnessCheck.readSafe(root, docPath).contains(heading)
                            }.map { heading ->
                                Finding(
                                    HarnessCheck.severityOf(manifest, category),
                                    category,
                                    "doc missing $heading: $docPath",
                                )
                            }
                    }
            }
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
